_pydevd_bundle/pydevd_xml.py

```diff
@@ -27,81 +27,133 @@ class ExceptionOnEvaluate:
     def __init__(self, result):
         self.result = result
 
-#------------------------------------------------------------------------------------------------------ resolvers in map
 
-_TYPE_MAP = None
 
 
-def _update_type_map():
-    global _TYPE_MAP
-    if not sys.platform.startswith("java"):
-        _TYPE_MAP = [
-                #None means that it should not be treated as a compound variable
 
-                #isintance does not accept a tuple on some versions of python, so, we must declare it expanded
-                (type(None), None,),
-                (int, None),
-                (float, None),
-                (complex, None),
-                (str, None),
-                (tuple, pydevd_resolver.tupleResolver),
-                (list, pydevd_resolver.tupleResolver),
-                (dict, pydevd_resolver.dictResolver),
-        ]
+class AbstractTypeResolveHandler(object):
+    __instance = None
 
-        try:
-            _TYPE_MAP.append((long, None))
-        except:
-            pass #not available on all python versions
+    def __init__(self,default_type_map):
+        super(AbstractTypeResolveHandler, self).__init__()
+        self.default_type_map = default_type_map
+        self.resolve_providers = pydevd_extension_utils.extensions_of_type(TypeResolveProvider)
+        self.raw_type_map = dict(self.default_type_map)
+        self.str_providers = pydevd_extension_utils.extensions_of_type(StrPresentationProvider)
 
-        try:
-            _TYPE_MAP.append((unicode, None))
-        except:
-            pass #not available on all python versions
 
+    def get_type(self,o):
         try:
-            _TYPE_MAP.append((set, pydevd_resolver.setResolver))
+            type_object = type(o)
+            type_name = type_object.__name__
         except:
-            pass #not available on all python versions
+            # This happens for org.python.core.InitModule
+            return 'Unable to get Type', 'Unable to get Type', None
 
-        try:
-            _TYPE_MAP.append((frozenset, pydevd_resolver.setResolver))
-        except:
-            pass #not available on all python versions
+        return self._get_type(o, type_object, type_name)
 
+    def _get_type(self, o, type_object, type_name):
+        # fast path most common types
+        if type_object in self.raw_type_map:
+            return type_object, type_name, self.raw_type_map[type_object]
         try:
-            import numpy
-            _TYPE_MAP.append((numpy.ndarray, pydevd_resolver.ndarrayResolver))
-        except:
-            pass  #numpy may not be installed
+            for provider in self.resolve_providers:
+                if provider.can_provide(type_object, type_name, o):
```

Comment:
Well, the version in this branch vs the version in my sample branch is around 10% slower, so, probably around the times you're having there (I think that 10% speed here is worth it -- I actually also plan to optimize this function as when stepping it's one of the slow parts of the process, probably having a custom quote/xml replace on cython when possible).

 Also, as you have more providers, things will probably get worse -- as for your sample, it's on the fastest path there... the worse would be always getting to a custom user object with no resolver (I'd also test without cprofile, just getting the time, as in this case the profile overhead is probably big).

Also note that experience may vary among the many python implementations (Jython in particular may have a runtime where times may not match what you see on CPython).